design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,162 @@
+3. lakeFS will need to track issued tokens/addresses, and delete them when tokens are expired/used
+4. LinkPhysicalAddress to verify token valid before creating an entry.
+    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
+       the link API and avoid them being deleted by the GC process.
+    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
+       eventually be deleted by the GC job.
+>**Note:** These changes also solves an existing bug where an issued physical address which was committed can be overwritten by Link
+
+#### S3 Gateway CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+
+#### Move/RenameObject
+Clients working through the S3 Gateway can use the CopyObject + DeleteObject to perform a Rename or Move operation.
+For clients using the OpenAPI this could have been done using StageObject + DeleteObject.
+To continue support of this operation, introduce a new API to rename an object which will be scoped to a single branch.
+
+#### PrepareUncommittedForGC
+
+A new API which will create meta-ranges and ranges for a given branch using its uncommitted data. These files
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `BranchUncommittedMetarange`
+
+### GC Flows
+
+The following describe the GC process run flows on a repository:
+
+#### Flow 1: Clean Run
+
+1. Listing namespace objects
+   1. Read all objects directly from object store (can be done in parallel using the partitions) -> `Store DF`
+   2. Skip partitions that are newer than < TOKEN_EXPIRY_TIME >
+2. Listing of lakeFS repository uncommitted objects
+   1. Mark uncommitted data
+      1. List branches
+      2. Run _PrepareUncommittedForGC_ on all branches
+   2. Get all uncommitted data addresses
+      1. For each branch
+         1. Read all addresses from branch `BranchUncommittedMetarange` -> `Uncommitted DF`
+    >**Note:** To avoid possible bug, uncommitted listing must happen before listing of committed data
+3. Listing of lakeFS repository committed objects
+   1. Get all committed data addresses
+      1. Read all addresses from Repository commits -> `Committed DF`
+4. Find candidates for deletion
+   1. Subtract committed data from all objects (`Store DF` - `Committed DF`)
+   2. Subtract uncommitted data from all objects (`Store DF` - `Uncommitted DF`)
+   3. Filter files in special paths
+   4. The remainder is a list of files which can be safely removed
+5. Save run information under the GC path
+    1. The current run's `Uncommitted DF` (as a parquet file)
+    2. The last read partition
+    3. The GC run start timestamp
```

Comment:
I think we need to structure the new objects in the GC prefix in the storage for alignment. Can be a followup task too.

Reply:
Done